<a href="https://colab.research.google.com/github/TurkuNLP/Deep_Learning_in_LangTech_course/blob/master/exercise_sentence_transformers_vs_xlmr_solved.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip3 install -q datasets transformers sentence-transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 519.6/519.6 kB 10.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.7/7.7 MB 80.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 10.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 kB 14.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 17.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 13.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.0/295.0 kB 31.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 99.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 76.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 80.5 MB/s eta 0:00:00


# English - MRPC

In [ ]:
import datasets

dataset_en = datasets.load_dataset("glue", "mrpc")

Generating train split:   0%|          | 0/3668 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/408 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1725 [00:00<?, ? examples/s]

In [ ]:
# Filter to positive cases to assure paraphrase pair exists
def is_paraphrase_pair(example):
    return example["label"] == 1

dataset_en = dataset_en.filter(is_paraphrase_pair)

Filter:   0%|          | 0/3668 [00:00<?, ? examples/s]

Filter:   0%|          | 0/408 [00:00<?, ? examples/s]

Filter:   0%|          | 0/1725 [00:00<?, ? examples/s]

In [ ]:
# Print the first 20 examples
for item in dataset_en["test"].select(range(20)):
    print(item)

{'sentence1': "PCCW 's chief operating officer , Mike Butcher , and Alex Arena , the chief financial officer , will report directly to Mr So .", 'sentence2': 'Current Chief Operating Officer Mike Butcher and Group Chief Financial Officer Alex Arena will report to So .', 'label': 1, 'idx': 0}
{'sentence1': "The world 's two largest automakers said their U.S. sales declined more than predicted last month as a late summer sales frenzy caused more of an industry backlash than expected .", 'sentence2': 'Domestic sales at both GM and No. 2 Ford Motor Co. declined more than predicted as a late summer sales frenzy prompted a larger-than-expected industry backlash .', 'label': 1, 'idx': 1}
{'sentence1': 'According to the federal Centers for Disease Control and Prevention ( news - web sites ) , there were 19 reported cases of measles in the United States in 2002 .', 'sentence2': 'The Centers for Disease Control and Prevention said there were 19 reported cases of measles in the United States in 2

In [ ]:
dataset_fi = datasets.load_dataset("TurkuNLP/turku_paraphrase_corpus", "plain")

Extracting data files:   0%|          | 0/4 [00:00<?, ?it/s]

Generating train split: 0 examples [00:00, ? examples/s]

Generating validation split: 0 examples [00:00, ? examples/s]

Generating test split: 0 examples [00:00, ? examples/s]

In [ ]:
# Filter to full paraphrases to assure paraphrase pair exists
def is_paraphrase_pair(example):
    return example["label"] == "4"

dataset_fi = dataset_fi.filter(is_paraphrase_pair)

Filter:   0%|          | 0/83645 [00:00<?, ? examples/s]

Filter:   0%|          | 0/10529 [00:00<?, ? examples/s]

Filter:   0%|          | 0/10471 [00:00<?, ? examples/s]

In [ ]:
# Print the first 20 examples
for item in dataset_fi["test"].select(range(20)):
    print(item)

{'id': 'turku_paraphrase_corpus-test-1', 'gem_id': 'gem-turku_paraphrase_corpus-test-1', 'goeswith': 'episode-11836', 'fold': 90, 'text1': 'Katsokaa hänen pikkuhampaita. Ihan kuin delfiinimies.', 'text2': 'Katsokaa hänen pikkuruisia hampaitaan, hän näyttää delfiinimieheltä.', 'label': '4', 'binary_label': 'positive', 'is_rewrite': True}
{'id': 'turku_paraphrase_corpus-test-3', 'gem_id': 'gem-turku_paraphrase_corpus-test-3', 'goeswith': 'episode-07366', 'fold': 90, 'text1': 'Tarkkailen tilanneta vielä muutaman minuutin. Sitten menen takaisin sisälle.', 'text2': 'Katson tilannetta vielä muutaman minuutin ajan ja menen uudelleen sisään.', 'label': '4', 'binary_label': 'positive', 'is_rewrite': True}
{'id': 'turku_paraphrase_corpus-test-5', 'gem_id': 'gem-turku_paraphrase_corpus-test-5', 'goeswith': 'episode-00231', 'fold': 90, 'text1': 'Et erotettuna jatka tämän jutun tutkimuksia.', 'text2': 'Sinut on erotettu tästä jutusta ja sen tutkimuksista.', 'label': '4', 'binary_label': 'positive',

# English vs Finnish

* Labels differ
* One has the texts in `sentence1` and `sentence2` and the other in `text1` and `text2` fields

# This is where the actual solution happens

* `xlm-roberta-base` is used
* the "feature-extraction" Pipeline() does what we need
* we just need to average the token embeddings, and normalize them so the rest of the code works

In [ ]:
## Old model loading code
from sentence_transformers import SentenceTransformer
model = SentenceTransformer('paraphrase-xlm-r-multilingual-v1')
model=model.cuda()

## A pipeline-based approach
from transformers import pipeline
import torch
pipe = pipeline("feature-extraction", model="xlm-roberta-base", device=0) #device=0 -> place on GPU


In [ ]:
# This is how one would do the embeddings
# with a pipeline

import torch

def embed_with_pipelines(texts,quiet=True):
    raw_emb=pipe(texts) #we avoid batching here, otherwise we would need to deal with padding
    #consequently, running this model is a bit slow
    averages=[] #accumulate the embedded sentences into a list, one at a time
    for one_sent_emb in raw_emb:
        one_sent_emb_tensor=torch.tensor(one_sent_emb)
        if not quiet:
            print("one_sent_emb shape",one_sent_emb_tensor.shape)
        #take the mean across the 2nd dimension and then get rid of the 1-long batch dimension
        averages.append(one_sent_emb_tensor.mean(1).squeeze()) #..acccumulate in a list
    # ...and vertically "stack" the list of embeddings, forming a single tensor
    embedded=torch.vstack(averages)
    print("Final shape:", embedded.shape)
    # we will yet need these normalized
    embedded=torch.nn.functional.normalize(embedded)
    return embedded
texts=["Hello there!","Hi my friend","There are no trams in Turku"]
embed_with_pipelines(texts,quiet=False)

one_sent_emb shape torch.Size([1, 5, 768])
one_sent_emb shape torch.Size([1, 5, 768])
one_sent_emb shape torch.Size([1, 9, 768])
Final shape: torch.Size([3, 768])


tensor([[ 0.0061,  0.0082, -0.0010,  ..., -0.0136, -0.0004,  0.0118],
        [ 0.0012,  0.0074,  0.0001,  ..., -0.0021,  0.0014,  0.0028],
        [ 0.0003,  0.0002,  0.0005,  ...,  0.0015,  0.0005,  0.0032]])

In [ ]:
#### UNCOMMENT FOR FINNISH

#dataset = dataset_fi # change to dataset_en for the English data
#text_field = "text"  # change to "sentence" for the English data
#max_count = 5000 # this is a huge dataset, let's take 5K examples

####  UNCOMMENT FOR ENGLISH

dataset = dataset_en # change to dataset_en for the English data
text_field = "sentence"  # change to "sentence" for the English data
max_count=len(dataset["train"]) #This is shorter dataset, we can take everything


texts1=[item[text_field+"1"] for item in dataset["train"].select(range(max_count))]
texts2=[item[text_field+"2"] for item in dataset["train"].select(range(max_count))]

In [ ]:
## Uncomment to embed with the SentenceTransformer
#encoded_texts1 = model.encode(texts1, convert_to_tensor=True, device=model.device, show_progress_bar=True, normalize_embeddings=True)
#encoded_texts2 = model.encode(texts2, convert_to_tensor=True, device=model.device, show_progress_bar=True, normalize_embeddings=True)

## Uncomment to embed with the vanilla XLMR
encoded_texts1=embed_with_pipelines(texts1)
encoded_texts2=embed_with_pipelines(texts2)


Final shape: torch.Size([2474, 768])
Final shape: torch.Size([2474, 768])


now we have the "left" sentences from the dataset encoded in `texts1_e` and the "right" sentences in `texts2_e`. Let us make sure:

In [ ]:
print(encoded_texts1.shape)
print(encoded_texts2.shape)

torch.Size([2474, 768])
torch.Size([2474, 768])


# Pairwise comparison

* We need to compare all sentences from `encoded_texts1` with all sentences from `encoded_texts2`
* We can use simple matrix multiplication to get all-against-all dot products, because that is exactly what matrix multiplication does
* Note that the embeddings are normalized (we asked for it in the `.encode()` arguments), so dot product is the same as cosine similarity
* This is efficient in the sense that it runs on GPU
* For large data which would not fit in the GPU memory, we'd need to do something else

![matrix multiplication](https://www.mcs.anl.gov/~itf/dbpp/text/img792.gif)

In [ ]:
import torch

sims=torch.mm(encoded_texts1, encoded_texts2.T)
sims.shape

torch.Size([2474, 2474])

## Most similar sentences

* We have done this many times previously
* To get the most similar pairs out of a matrix of similarities, we do `argsort()` or `argmax()` and pick the first one

In [ ]:
sims_sort = torch.argsort(sims, dim=-1, descending=True)

* stop to think what you would expect the result to look like
* we compare the similarity of "left" sentences to "right" sentences

In [ ]:
sims_sort[:100,0]

tensor([   0,    1, 1290,  136,    4,    5,    6,    7,    8,    9,   10,   11,
          12,   13,   14,   15, 1710,   17,   18, 2336,   20,   21,   22,   23,
          24,   25,   26,   27,   28, 1319,   30,   31, 1903,   33,   34,   35,
          36,   37,   38,   39,   40,   41, 1787,   43,   44,   45,   46,   47,
          48,   49,   50,   51, 1849,   53,  834, 1276,   56,   57,   58,   59,
          60,   61,   62,   63,   64,   65,   66,   67,   68,   69, 1284,   71,
          72,   73,   74,   75,   76,   77,   78, 2165,   80,   81,   82, 1907,
          84,   85,   86,   87,   88,   89,   90,   91,   92,   93,   94,  133,
          96,   97,  512,   99])

In [ ]:
# let's inspect few pairs
for i in range(5):
    print(dataset["train"][i][text_field+"1"])
    j=int(sims_sort[i,0]) #this is the index of the corresponding sentence
    print(dataset["train"][j][text_field+"2"]) #
    print("\n---------------------\n")


Amrozi accused his brother , whom he called " the witness " , of deliberately distorting his evidence .
Referring to him as only " the witness " , Amrozi accused his brother of deliberately distorting his evidence .

---------------------

They had published an advertisement on the Internet on June 10 , offering the cargo for sale , he added .
On June 10 , the ship 's owners had published an advertisement on the Internet , offering the explosives for sale .

---------------------

The stock rose $ 2.11 , or about 11 percent , to close Friday at $ 21.51 on the New York Stock Exchange .
IBM shares closed up $ 1.75 , or 2.11 percent , at $ 84.50 on the New York Stock Exchange .

---------------------

Revenue in the first quarter of the year dropped 15 percent from the same period a year earlier .
In the first quarter of this year , weekly earnings adjusted for inflation fell 1.5 percent - the biggest drop in more than a decade .

---------------------

The DVD-CCA then appealed to the st

# Quick evaluation on accuracy

In [ ]:
# just a very simple way to calculate accuracy
# Perfect answer should be 0,1,2,3,4... the way our data works
correct=0
for i in range(sims_sort.shape[0]):
    if sims_sort[i,0]==i:
        correct+=1
print(f"Accuracy: {correct/sims_sort.shape[0]*100}%")

Accuracy: 82.82134195634599%


# Results:

* English XLMR: 82.8%
* English SentenceTransformer: 90.78%
* Finnish XLMR: 60.58%
* Finnish SentenceTransformer: 90.28%

SentenceTransformers do work better, but the difference for Finnish is much higher than for English.

When checking out the data, it seems the English data is relatively simple with huge lexical overlap. The Finnish data is much harder, striving to avoid lexical overlap (we know, we made the data :)